client/app/scripts/superdesk/editor/editor.js

```diff
@@ -550,16 +550,375 @@ function EditorService(spellcheck, $rootScope, $timeout) {
     }
 }
 
-angular.module('superdesk.editor', ['superdesk.editor.spellcheck'])
+angular.module('superdesk.editor', ['superdesk.editor.spellcheck', 'angular-embed'])
 
     .service('editor', EditorService)
-
-    .directive('sdTextEditor', ['editor', 'spellcheck', '$timeout', function (editor, spellcheck, $timeout) {
-
-        var config = {
-            buttons: ['bold', 'italic', 'underline', 'quote', 'anchor'],
-            anchorInputPlaceholder: gettext('Paste or type a full link'),
-            disablePlaceholders: true,
+    .directive('sdAddEmbed', function() {
+        return {
+            scope: true,
+            require: ['sdAddEmbed', '^sdTextEditor'],
+            templateUrl: 'scripts/superdesk/editor/views/add-embed.html',
+            controllerAs: 'vm',
+            controller: ['embedService', '$element', '$timeout', '$q', 'lodash',
+                        function(embedService, $element, $timeout, $q, _) {
+                var vm = this;
+                angular.extend(vm, {
+                    blockBefore: undefined,  // defined in link method
+                    editorCtrl: undefined,  // defined in link method
+                    extended: false,
+                    toggle: function(close) {
+                        // use parameter or toggle
+                        vm.extended = angular.isDefined(close) ? !close : !vm.extended;
+                        // on enter, focus on input
+                        if (vm.extended) {
+                            $timeout(function() {
+                                angular.element($element).find('input').focus();
+                            });
+                        // on leave, clear field
+                        } else {
+                            vm.input = '';
+                        }
+                    },
+                    retrieveEmbed:function() {
+                        // if it's an url, use embedService to retrieve the embed code
+                        var embedCode;
+                        if (_.startsWith(vm.input, 'http')) {
+                            embedCode = embedService.get(vm.input).then(function(data) {
+                                return {
+                                    body: data.html,
+                                    provider: data.provider_name
+                                };
+                            });
+                        // otherwise we use the content of the field directly
+                        } else {
+                            var embedType = 'Custom';
+                            // try to guess the provider of the custom embed
+                            if (vm.input.indexOf('twitter.com/widgets.js')) {
+                                embedType = 'Twitter';
+                            } else if (vm.input.indexOf('https://www.youtube.com')){
+                                embedType = 'YouTube';
```

Comment:
It might make sense to define these constant strings in some sort of an "enum", easier to reuse them at other places if needed.